src/mame/drivers/uknc.cpp

```diff
@@ -9,97 +9,946 @@ Graphics 640x288 pixels.
+		m_maincpu->resume(SUSPEND_REASON_DISABLE);
+	}
+	else if (!BIT(data, 15) && !m_maincpu->suspended(SUSPEND_REASON_DISABLE))
+	{
+		m_maincpu->set_input_line(INPUT_LINE_ACLO, ASSERT_LINE);
+	}
+#endif
+}
+
+/*
+ *	0-3	CE0..CE3 bus signals
+ *	  CE0	mapping at 100000: 0 - read from VRAM plane 0, 1 - read from ROM (reset to 1)
+ *	  CE1-2	ROM carts: bank selector
+ *	  CE3	slot number
+ *	4-7	ROM banks replacement: 0 - writes ignored, 1 - writes go to VRAM plane 0
+ *	  4		100000..117777
+ *	  5		120000..137777
+ *	  6		140000..157777
+ *	  7		160000..177777
+ *	8	line clock interrupt (subcpu): 0 - enable
+ *	9	line clock interrupt (maincpu): 0 - enable
+ */
+uint16_t uknc_state::bcsr_r()
+{
+	return m_bcsr;
+}
+
+void uknc_state::bcsr_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	int bank1, bank2;
+	address_space *subspace = &m_subcpu->space(AS_PROGRAM);
+
+	bank1 = (data & 1) | ((data >> 3) & 1);
+	bank2 = (data >> 5) & 7;
+
+	// order matters: ROM on device may unmap itself and we restore the mapping
+//	m_subqbus->ce0_ce3_w(data & 15);
+
+	if (BIT(m_bcsr ^ data, 0) && BIT(data, 0))
+	{
+		subspace->unmap_readwrite(0100000, 0117777);
+		subspace->install_readwrite_handler(0100000, 0117777,
+			read16s_delegate(m_bankdev1, FUNC(address_map_bank_device::read16)),
+			write16s_delegate(m_bankdev1, FUNC(address_map_bank_device::write16)));
+	}
+
+	m_bankdev1->set_bank(bank1);
+	m_bankdev2->set_bank(bank2);
+
+	m_bcsr = data;
+}
+
+// video
+
+#define YRGB(rgb, y, p) ((rgb)*42 * (3 + (p) + (y)*2))
+
+void uknc_state::set_palette_yrgb(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 2), BIT(aux, i * 4 + 3), BIT(m_video.control, 18)),
+			YRGB(BIT(aux, i * 4 + 1), BIT(aux, i * 4 + 3), BIT(m_video.control, 17)),
+			YRGB(BIT(aux, i * 4    ), BIT(aux, i * 4 + 3), BIT(m_video.control, 16)))
+		);
+	}
+}
+
+void uknc_state::set_palette_ygrb(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 1), BIT(aux, i * 4 + 3), BIT(m_video.control, 17)),
+			YRGB(BIT(aux, i * 4 + 2), BIT(aux, i * 4 + 3), BIT(m_video.control, 18)),
+			YRGB(BIT(aux, i * 4    ), BIT(aux, i * 4 + 3), BIT(m_video.control, 16)))
+		);
+	}
+}
+
+void uknc_state::set_palette__rgb(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 2), !BIT(aux, i * 4 + 3), 1),
+			YRGB(BIT(aux, i * 4 + 1), !BIT(aux, i * 4 + 3), 1),
+			YRGB(BIT(aux, i * 4    ), !BIT(aux, i * 4 + 3), 1))
+		);
+	}
+}
+
+void uknc_state::set_palette_gray(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(i * 255 / 8, i * 255 / 8, i * 255 / 8));
+	}
+}
+
+#undef YGRB
+
+
+void uknc_state::update_displaylist()
+{
+	int row, wide = 0;
+	uint16_t addr = 0270, nextaddr;
+	bool cursor = false, cursor_type = false;
+	int cursor_color = 0, cursor_octet = 0, cursor_pixel = 0;
+	scanline_t *scanline;
```

Comment:
You should scope these variables appropriately, and use `const` where it makes sense.

Reply:
changed.